BIP-325: Signet support

kallewoof opened the pull request:
This PR implements BIP 325 (https://github.com/bitcoin/bips/blob/master/bip-0325.mediawiki) -- Signet support.

**Note:** This PR is being split into more bite-sized PR's, the first of which is #18267.

Merged check boxes:
* [ ] Signet consensus: #18267
* [ ] Signet RPC tools: pending consensus merge
* [ ] Signet utility scripts (contrib/signet): pending RPC tool merge

Signet is a proposed new type of test network that requires a signature in the blocks. While this kind of network is centralized, it allows for properties that are desirable in a test network, which testnet alone cannot provide.

Signet has a default global testnet baked in for easy usage, but is built around the concept of there being multiple simultaneous signets. Someone is already working on a signet with bip-taproot etc patched on top of it. When new features are tested globally, starting a temporary signet for that purpose is trivial.

See also: https://en.bitcoin.it/wiki/Signet

Noteworthy:
* ~~proof of work check now skips genesis block for all networks, including mainnet~~
* a new global "is signet" flag is added, which affects consensus: it adds a new enforcement that all blocks must have a valid signature for the "signet block script", a global script
* `DecodeHexBlk` will now attempt to load the block data from a file whose filename is the hex string argument, if it turns out to not be a valid hex string; this has the odd caveat of the relative directory being relative to the execution of `bitcoind`, not `bitcoin-cli` (see `contrib/signet/mkblock.sh`)

Update: proof of work check is back in place; instead, Signet now requires an additional `-signet_genesisnonce` parameter to operate.

kallewoof committed:
signet: hard-coded parameters for Signet Global Network IV (2020-03-05)

Running with -signet without any other parameters will use this network's parameters.

kallewoof committed:
contrib: add signet scripts (issuer, etc)

kallewoof committed:
add witness commitment section support to primitive/block

This also moves GetWitnessCommitmentIndex out of validation.cpp into CBlock, with 3 lines affected (GetWitnessCommitmentIndex(block) -> block.GetWitnessCommitmentIndex()).

kallewoof committed:
add signet basic support (signet.cpp)

kallewoof committed:
qt: update QT to support signet network

kallewoof committed:
consensus: add signet validation

kallewoof committed:
rpc: add signblock command

kallewoof committed:
rpc: include signet commitment in blockToJSON dictionary results

kallewoof committed:
rpc: add getnewblockhex command for obtaining an unsigned block

kallewoof committed:
i/o: make DecodeHexBlk able to read block hex from file

kallewoof commented:
@practicalswift Would you mind helping me figure out how to make linters happy? https://travis-ci.org/bitcoin/bitcoin/jobs/560317378

junderw commented:
Brainstorming some things we should consider:

1. Does adding this put anyone at risk?
2. Does adding this somehow hurt Bitcoin's code base?
3. Does adding this require some long term maintenance / increase the weight on devs for support in the future?

1. Since Testnet has been used at some points to scam people out of money, I could see how this could be used in a similar way. However, with something like this it would be easier for someone to create a signet and give it value (aka use it to create a permission-ed altcoin and mooch 100% off Bitcoin Core development. At least with current altcoin, they have a "keep up to date with merging Core patches" barrier of entry... this would not have such a barrier.)
2. I would say no. But that would need more in-depth technical review.
3. If scams run rampant in my scenario in number 1, some people might propose hard coding blacklists that will require scammers to maintain patches removing themselves from the signet blacklist etc... But other than that, I don't see much problem.

I am leaning towards concept ACK. But I am slightly concerned about someone abusing this like testnet has been abused in the past, and how signet's differences make those abuses easier or harder.

kallewoof commented:
@junderw

Thanks for feedback, comments below:

>     1. Does adding this put anyone at risk?

Example?

>     2. Does adding this somehow hurt Bitcoin's code base?

There is a tiny amount of risk involved as it touches consensus validation code, but this is a tiny code change.

If for some reason someone managed to flip the "signet" flag on a running node, for example, that would cause those nodes to start rejecting bitcoin blocks. I don't foresee this as very likely, personally.

>     3. Does adding this require some long term maintenance / increase the weight on devs for support in the future?

Yes, same as adding any feature to a system. I assume the point of your question is if the added maintenance is worth it; in my opinion it is. Signet is fairly self-contained (3 main parts: signet, wallet RPC, miner RPC, and some sprinkled stuff; there are also the contrib/signet scripts but I see those as irrelevant in this case as they're not a part of the C++ codebase), so it shouldn't be that hard to maintain.

>     4. Since Testnet has been used at some points to scam people out of money, I could see how this could be used in a similar way. However, with something like this it would be easier for someone to create a signet and give it value (aka use it to create a permission-ed altcoin and mooch 100% off Bitcoin Core development. At least with current altcoin, they have a "keep up to date with merging Core patches" barrier of entry... this would not have such a barrier.)

I don't think this is an issue, to be honest. People can already copy the codebase and tweak it to make a new coin, or start up a regtest and ask their victim to connect to it using This Special Bitcoin.conf File [tm] which so happens to have regtest=1 in it.

junderw commented:
> People can already copy the codebase and tweak it to make a new coin, or start up a regtest and ask their victim to connect to it using This Special Bitcoin.conf File [tm] which so happens to have regtest=1 in it.

Which is why I am mostly leaning toward concept ACK.

But tbh, if someone said "hey, we have a 7-of-13 multisig based faster Bitcoin!" as compared to "hey use this regtest thing, totally decentralized, we swear!" could be argued by pedantic people as "not technically false"... so a regtest based scam has less of a chance of someone understanding the tech aspect saying "yeah, that's not centralized" than a signet based scam.

Also, there are also many scams that copy-pasted the code base, sure... but with a signet based scam, they don't even need to maintain a patch. Just tell the victims to download Bitcoin Core. And it works out of the box (with a conf tweak)...

So while the risk is super low, it is objectively higher than regtest and testnet...

That being said. The benefit is great compared to that risk. Which, again, is why I say "leaning towards concept ACK."

It would be very powerful to have someone who already has Bitcoin Core downloaded and running to spin up another instance with some args that lets them try out taproot etc etc...

kallewoof commented:
Yep, I see what you're saying. Ultimately both cases require a conf tweak, though (regtest or testnet scam vs signet scam), and signet case requires them to give you a big ugly signet challenge and signet seed node(s), so I'm not sure how viable the attack really is in the end.

NicolasDorier commented:
Concept ACK, though I am unsure if signet should be configurable at all. (via `signet_blockscript` for example)

It should just be a shared public network, an alternative (or replacement) of testnet.

NicolasDorier commented:
The code does not seems to be that much to maintain, I would advise though to open a separate pull requests for most changes in `validation.cpp/h`.

NicolasDorier commented on src/validation.cpp:
```diff
@@ -941,9 +941,12 @@ bool ReadBlockFromDisk(CBlock& block, const FlatFilePos& pos, const Consensus::P
-    if (!CheckProofOfWork(block.GetHash(), block.nBits, consensusParams))
-        return error("ReadBlockFromDisk: Errors in block header at %s", pos.ToString());
+    // We skip POW checks for genesis block
+    if (block.GetHash() != consensusParams.hashGenesisBlock) {
```
I would move this test inside `CheckProofOfWork`

kallewoof replied:
I need this if block for `CheckBlockSolution` later, so I would have to do the if case anyway.

NicolasDorier commented on src/validation.cpp:
```diff
@@ -2965,10 +2968,13 @@ bool CheckBlock(const CBlock& block, CValidationState& state, const Consensus::P
-    if (!CheckBlockHeader(block, state, consensusParams, fCheckPOW))
-        return false;
+    // We skip POW checks for genesis block
+    if (block.GetHash() != consensusParams.hashGenesisBlock) {
```
I would move this test inside `CheckBlockHeader`

kallewoof replied:
I need this if block for `CheckBlockSolution` too, though.

NicolasDorier replied:
`CheckBlockSolution` could also return true if  `block.GetHash() == consensusParams.hashGenesisBlock`

kallewoof replied:
That's fair. But the check appears in 3 places instead of 2. Not sure why that is better, tbh.

MarcoFalke commented:
Would be nice to have a `contrib/signet/README.md` to explain what those sh scripts are supposed to be doing

kallewoof commented:
@MarcoFalke Added README file. Also missing release notes. Since it is Needs Concept Review stage, I am not gonna rush that just yet.

kallewoof commented:
I added a "noteworthy" section to the OP of this PR, which points out a number of changes to existing functionality.

practicalswift commented:
Concept ACK

Thanks for doing this - would be very useful for testing.

MarcoFalke commented on contrib/signet/README.md:
```diff
@@ -0,0 +1,60 @@
+
+A script to generate one Signet block.
+
+Syntax: `mkblock.sh <bitcoin-cli path> [<bitcoin-cli args>]`
```
How is this supposed to work? It looks like the chainparams are setting the pow to the pow of mainnet. This script calls `grindblock`, which is a non-optimized while loop on a single CPU core. Could this lead to the chain stalling?

junderw replied:
If the signers compete with each other, or frequently change their hashpower, yes.

I think the idea is that signatures limit the miner pool to a few people, then those people each pledge a consistent low hashpower to the network.

Prevents people from hopping on with ASICs, driving up difficulty and then disappearing.

kallewoof replied:
The pow is not really the pow of mainnet unless you don't pass `-signet` or use a `-datadir` with signet in bitcoin.conf.

`grindblock` is basically equivalent to `generate*`. You can use other miners like existing CPU miners or even ASICs (though you can't use the extra nonce so you'd have to resign the block instead).

kallewoof committed:
test: add signet mining tests

Also adds a condition to test initialization code to not generate blocks for signet chains.

kallewoof commented:
Refactored now that #16509 is in (one less commit, and fewer (test) code lines touched!).

michaelfolkson commented:
Is there a resource online discussing the difference(s) between regtest and (proposed) signet? I found this on the [bitcoin-dev mailing list](https://lists.linuxfoundation.org/pipermail/bitcoin-dev/2019-March/016742.html) but it doesn't go into much detail. Is unvetted "people rewriting the blockchain at their whim by mining a ton of blocks" a common problem? It wouldn't be a problem for a training workshop (just spin up a new one) but could possibly be for a project building on top of regtest.

Also any other key differences between design of regtest and signet? I don't need convincing of signet's upsides in comparison to testnet but not convinced re regtest versus signet. Digging into understanding regtest code a bit more but any guidance would be appreciated.

kallewoof commented:
@michaelfolkson I already replied on Twitter, but there is a [Bitcoin Wiki entry](https://en.bitcoin.it/wiki/Signet) and also a [Bitcoin Stack Exchange page](https://bitcoin.stackexchange.com/questions/89640/what-are-the-key-differences-between-regtest-and-the-proposed-signet?stw=2) (made by you).

kallewoof commented:
The proof of work changes were removed; Signet now requires a POW-valid genesis block, which is generated using `grindblock <challenge>` and which is provided to nodes using `-signet_genesisnonce`.

NicolasDorier commented:
> Update: proof of work check is back in place; instead, Signet now requires an additional `-signet_genesisnonce`

It should not require. If I want to use signet, I want to do `bitcoind -signet` without any more obscure parameter.

kallewoof commented:
@NicolasDorier The default parameters includes a genesis nonce value. It's only for custom signets, where you provide your own `-signet_blockscript` as well.

NicolasDorier commented on src/chainparams.cpp:
```diff
@@ -7,6 +7,8 @@
 #include <chainparamsseeds.h>
 #include <consensus/merkle.h>
+#include <hash.h> // for signet block challenge hash
+#include <signet.h>
```
I think you are not using this one in this file.

kallewoof replied:
I do; I reference `g_signet_blocks`, for example.

jtimon replied:
I think it's only for that, right? Perhaps a small reason to put the global here instead of there.

kallewoof replied:
Belated, but I am removing this and adding an `extern CScript g_signet_blockscript` in the signet chainparams class.

NicolasDorier commented on src/core_read.cpp:
```diff
@@ -158,10 +158,27 @@ bool DecodeHexBlockHeader(CBlockHeader& header, const std::string& hex_header)
-    if (!IsHex(strHexBlk))
+    if (!IsHex(strHexBlk)) {
+        std::string actual;
+        if (ReadFromDisk(strHexBlk, actual)) return DecodeHexBlk(block, actual);
```
why do you need this?

kallewoof replied:
Juggling block hex values in Bash is very fragile. If a block gets too big, bash scripts start breaking. To avoid, the issuer script redirects output to files during the signing/grinding process.

NicolasDorier commented on src/rpc/mining.cpp:
```diff
@@ -171,6 +183,10 @@ static UniValue generatetoaddress(const JSONRPCRequest& request)
     }
 
+    if (g_signet_blocks) {
+        throw JSONRPCError(RPC_INVALID_PARAMETER, "Generating signet blocks require that you use getnewblockhex");
```
is there really no way to make it works?

kallewoof replied:
Recently, the wallet code and the rest of the code have been separated. Unfortunately this means the mining part and the signing part have to be separated as well.

NicolasDorier replied:
I don't understand. Why are you talking about wallet code? `generatetoaddress` does touch only mining code.

kallewoof replied:
I think we discussed this outside of here, but basically, you don't have the private keys associated with any addresses, so you'd have to provide those every time here. I don't think we should encourage systems that behave in that manner ("bitcoin core does it so it must be safe").

NicolasDorier commented on src/script/sign.h:
```diff
@@ -32,6 +32,16 @@ class BaseSignatureCreator {
     virtual bool CreateSig(const SigningProvider& provider, std::vector<unsigned char>& vchSig, const CKeyID& keyid, const CScript& scriptCode, SigVersion sigversion) const =0;
 };
 
+class SimpleSignatureCreator : public BaseSignatureCreator
```
comments about the purpose of this would be nice

kallewoof replied:
Added a few words to describe what it is. :)

NicolasDorier commented:
So I discussed with @kallewoof I don't like the fact that `generatetoaddress` does not work on custom signets.
The reason why `generatetoaddress` does not work on signet is that signet need blocks to be signed, and right now it uses the bitcoin wallet to find the keys. However `generatetoaddress` can't depends on wallet code.

Now we have two ways of fixing the problem:
1. Passing the keys for signing on signet via command line `--signet_privatekey` (My preference)
2. Passing the keys to `generatetoaddress` (@kallewoof preference)

I don't like 2, because it does not solve my initial problem: to have a custom signet just work with the same code you would use for regtest.

@kallewoof does not like my preference, because he says passing private keys via command line is bad practice. Personally, I don't think this is a problem because custom signet coins have no value, so no damage can be done.

Sjors commented:
@NicolasDorier or create `generatetodescriptor` which uses the block height as the index.

practicalswift commented:
@NicolasDorier @kallewoof Is the risk scenario that the signet private key would leak via process listings? If that it deemed a problem then it could perhaps be passed as an environment variable?

NicolasDorier commented:
@Sjors I don't really understand what you mean, but anyway there should not have "special way of doing same thing" in RPC depending if you are on custom signet or regtest.

@practicalswift it feel a bit weird to use environment variable when everything else is not.
I don't think that the key leaking is in anyway a problem that need to be solved.

Sjors commented:
@NicolasDorier instead of providing a single address or private key, you provide a [descriptor](https://github.com/bitcoin/bitcoin/blob/master/doc/descriptors.md) so that each block will have a unique payout address. E.g. if you pass `wpkh(xpub/0/*)` it could replace `*` with the block height. You can use `importmulti` with the private key `wpkh(xpriv/0/*)` to import that descriptor in any wallet, from which you can then spend the coins.

practicalswift commented:
> it feel a bit weird to use environment variable when everything else is not.

Yes, that's a good point.

Perhaps I'm missing something but wouldn't setting `signet_privatekey` in the config file solve the key-leak-via-process-listing issue (assuming it is an issue)?

kallewoof commented:
I don't agree with the premise that the RPC `generatetoaddress` must be possible under Signet. If you are generating the blocks yourself, use regtest, or use the Signet solution available. Signet is for more global testing where you often are not the miner.

I'm neutral on the private key juggling, as long as we don't teach people bad habits for valuable private keys.

NicolasDorier commented:
@Sjors I see, well my complaint is generally about, as a RPC user, I expect that RPC works as it would with regtest/testnet/mainnet. If I need to make an if/else in my code to specifically handle the case of a custom signet this would really suck.

jtimon commented on src/validation.cpp:
```diff
@@ -968,6 +969,8 @@ bool ReadBlockFromDisk(CBlock& block, const FlatFilePos& pos, const Consensus::P
     if (!CheckProofOfWork(block.GetHash(), block.nBits, consensusParams))
         return error("ReadBlockFromDisk: Errors in block header at %s", pos.ToString());
 
+    if (g_signet_blocks && block.GetHash() != consensusParams.hashGenesisBlock && !CheckBlockSolution(block, consensusParams)) return false; /* function calls error(..) */
```
instead of g_signet_blocks we could use consensusParams.signet_blocks or similar.

kallewoof replied:
Did this, but did not move `blockscript` - see detached comment.

jtimon commented on src/validation.cpp:
```diff
@@ -3036,6 +3039,8 @@ bool CheckBlock(const CBlock& block, CValidationState& state, const Consensus::P
         return false;
 
+    if (g_signet_blocks && fCheckPOW && block.GetHash() != consensusParams.hashGenesisBlock && !CheckBlockSolution(block, consensusParams)) return false;
+
```
Perhaps do this in CheckBlockHeader?
That way unsigned blocks would be detected faster.
Also, I think the return false in a new line would be more readable.

kallewoof replied:
`CheckBlockSolution` requires the block, not just the header.

jtimon commented on src/signet.cpp:
```diff
@@ -0,0 +1,58 @@
+#include <util/system.h>
+
+bool g_signet_blocks = false;
+CScript g_signet_blockscript;
```
Likewise this can be in consensus params instead of being a new global.

kallewoof replied:
It can, but it would be wasteful; see detached comment.

jtimon commented on src/rpc/blockchain.cpp:
```diff
@@ -160,6 +161,12 @@ UniValue blockToJSON(const CBlock& block, const CBlockIndex* tip, const CBlockIn
+    if (g_signet_blocks) {
+        std::vector<uint8_t> signet_commitment;
+        if (block.GetWitnessCommitmentSection(SIGNET_HEADER, signet_commitment)) {
+            result.pushKV("signet-solution", HexStr(signet_commitment));
```
it would be nice to also show the challenge in getblockchaininfo.

kallewoof replied:
True! Adding.

jtimon reviewed (commented):
I still want to do more review and testing, but left some comments already.

kallewoof committed:
rpc/signet: show signet blockscript in 'getblockchaininfo'

kallewoof commented:
Moved `g_signet_blocks` bool into consensus parameters.

However, moving `g_signet_blockscript` would require the `params.h` file to now `#include <script/script.h>` to get the `CScript`, which seems wasteful.

Alternatively I could make it a `std::vector<unsigned int>` in the consensus params, but then every time it's used, it would have to be converted, so you'd get `CScript(consensus.signet_blockscript.begin(), consensus.signet_blockscript.end())` sprinkled all over which doesn't seem ideal.

jtimon commented:
I think including script in consensus/params.h should be fine. I'm not sure
I understand "wasteful" in this context. Is going to "waste" exactly the
same memory.

But, if not, can we at least put it in chainparams?

On Fri, Aug 23, 2019, 07:18 kallewoof <notifications@github.com> wrote:

> Moved g_signet_blocks bool into consensus parameters.
>
> However, moving g_signet_blockscript would require the params.h file to
> now #include <script/script.h> to get the CScript, which seems wasteful.
>
> Alternatively I could make it a std::vector<unsigned int> in the
> consensus params, but then every time it's used, it would have to be
> converted, so you'd get CScript(consensus.signet_blockscript.begin(),
> consensus.signet_blockscript.end()) sprinkled all over which doesn't seem
> ideal.
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/bitcoin/bitcoin/pull/16411?email_source=notifications&email_token=AAHWGSV4TRS4YXT62FMOW4TQF5XLJA5CNFSM4IEW75LKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOD47EPWA#issuecomment-524175320>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAHWGSSCWTNGVH6XKSSEVXDQF5XLJANCNFSM4IEW75LA>
> .
>

kallewoof commented:
@jtimon 

> But, if not, can we at least put it in chainparams?

I mean, yes, we could totally do that, but I'm not sure which is better here to be honest. Putting it in chainparams means it's close to the other stuff, which is good. But keeping it in signet.h means isolating signet stuff more from the rest of the code, which is *also* good.

jtimon commented on src/rpc/mining.cpp:
```diff
@@ -948,6 +964,107 @@ static UniValue estimaterawfee(const JSONRPCRequest& request)
+        }
+    } else {
+        auto bin = ParseHexV(request.params[0], "challenge");
+        block = CreateSignetGenesisBlock(CScript(bin.begin(), bin.end()));
```
Why do we need to mine the genesis block, isn't it easier to just not validate the genesis block beyond its hash defined in chainparams?

kallewoof replied:
https://github.com/bitcoin/bitcoin/pull/16630#issuecomment-521988316

jtimon replied:
With https://github.com/bitcoin/bitcoin/pull/17037 we wouldn't need to grind the genesis block, and it is smaller than #8994 .

jtimon commented:
I think it's more consistent in chainparams, but I can see your point. Both options are valid, that's just what I prefer.

jtimon commented on src/chainparams.cpp:
```diff
@@ -265,20 +265,27 @@ class SigNetParams : public CChainParams {
+            vSeeds.push_back("2a01:7c8:d005:390::5");
+            vSeeds.push_back("ntv3mtqw5wt63red.onion:38333");
+        } else {
+            if (!args.IsArgSet("-signet_blockscript")) {
```
This will never trigger as it has already been checked and handled. It can be removed.

jtimon commented on src/chainparams.cpp:
```diff
@@ -265,20 +265,27 @@ class SigNetParams : public CChainParams {
+            if (!args.IsArgSet("-signet_blockscript")) {
+                throw std::runtime_error(strprintf("%s: -signet_blockscript is mandatory for signet networks", __func__));
+            }
+            if (args.GetArgs("-signet_blockscript").size() != 1) {
```
Why not just get the arg as a string?

kallewoof replied:
I figured it was good to check there weren't multiple blockscripts provided.

jtimon commented on src/chainparams.cpp:
```diff
@@ -246,6 +254,89 @@ class CTestNetParams : public CChainParams {
+        LogPrintf("SigNet with block script %s\n", gArgs.GetArgs("-signet_blockscript")[0]);
+
+        strNetworkID = "signet";
+        consensus.signet_blocks = true;
```
Can you set it to false in the others like regtest does with fPowNoRetargeting ?

kallewoof replied:
OK!

jtimon commented:
Created https://github.com/kallewoof/bitcoin/pull/9 with code solving most nits if not all.
Also created https://github.com/jtimon/bitcoin/pull/12 to see how signet and custom chains could be combined best.
Comments welcomed on any of them, and hopefully you will pick things from 9.
The last commit requires a restart of the default network for it changes its genesis block.

kallewoof committed:
util_tests.cpp update hash for hard coded chain merge

kallewoof commented:
@jtimon Sorry for lack of responsiveness. I've addressed your nits separately and will compare to your branch.

MarcoFalke commented on src/chainparamsbase.cpp:
```diff
@@ -23,6 +24,11 @@ void SetupChainParamsBaseOptions()
     gArgs.AddArg("-segwitheight=<n>", "Set the activation height of segwit. -1 to disable. (regtest-only)", ArgsManager::ALLOW_ANY | ArgsManager::DEBUG_ONLY, OptionsCategory::DEBUG_TEST);
     gArgs.AddArg("-testnet", "Use the test chain. Equivalent to -chain=test.", ArgsManager::ALLOW_ANY, OptionsCategory::CHAINPARAMS);
     gArgs.AddArg("-vbparams=deployment:start:end", "Use given start/end times for specified version bits deployment (regtest-only)", ArgsManager::ALLOW_ANY | ArgsManager::DEBUG_ONLY, OptionsCategory::CHAINPARAMS);
+    gArgs.AddArg("-signet", "Use the signet chain. Note that the network is defined by the signet_blockscript parameter", ArgsManager::ALLOW_ANY, OptionsCategory::CHAINPARAMS);
```
in commit 93dd0a1e62b019a72657f2abd60a2afc3d0702ba:

Can be removed, as this is redundant to `-chain=signet`

kallewoof replied:
Isn't it good to be able to do just -signet?

jtimon commented:
No worries, most of the commits in the branch are fixes to my own nits any way in case you find it convenient to squash them or something.

Regarding https://github.com/bitcoin/bitcoin/pull/16630#issuecomment-521988316 
@MarcoFalke I strongly disagree. the only reason we need to have rpc functionality to grind genesis blocks and the only reason this patch needs the signet_nonce parameter is validating the pow in the genesis block.

There's little reason to validate anything in the genesis block, for arguably the blocks is valid by definition and it's itself a consensus rule (the sencond block must contain its hash). But philosophical questions aside, the changes in the consensus code are quite simple and innocuous.
Not the way they're written in https://github.com/bitcoin/bitcoin/pull/16630 though, for it has a performance penalty for other valid blocks. The way it is written in https://github.com/bitcoin/bitcoin/pull/9102 there's no performance penalty at all unless the block is invalid, which I think it's fine, since the penalty should be small anyway, I haven't benchmarked it but it's just comparing 2 hashes right before logging an error.

Sorry for the rant, but I really dislike the -signet_nonce argument.
See https://github.com/kallewoof/bitcoin/pull/9/commits/c0063a5998061c524ee53c7fd80f83adaf01770e

jtimon commented on src/chainparams.cpp:
```diff
@@ -38,6 +40,12 @@ static CBlock CreateGenesisBlock(const char* pszTimestamp, const CScript& genesi
     return genesis;
 }
 
+static CBlock CreateGenesisBlock(const char* pszTimestamp, const CScript& genesisOutputScript, uint32_t nTime, uint32_t nNonce, uint32_t nBits, int32_t nVersion, const CAmount& genesisReward)
```
This function is only called from the next one and can be inlined. The change in the previous one is, I think all you need for your own CreateSignetGenesisBlock, which I would also prefer at the top (or at least before SigNetParams) in case it  turns into a static function in the future.
Really small nits.

kallewoof replied:
I don't think `CreateSignetGenesisBlock` will be static anytime soon, since it is also used in `mining.cpp` but I can definitely inline this one.

jtimon replied:
Well, it can only be static if we stop supporting grinding the genesis block because the genesis block's pow is not checked anymore and we don't need it.
So I guess it depends on the other discussion. If we end up doing that, we won't need to move the function later. But it's a very minor nit, feel free to ignore.

kallewoof replied:
Good point. If we end up reversing the grinding approach I will try to remember to address this.

kallewoof replied:
One part about skipping genesis PoW that shouldn't be forgotten is that this doesn't just mean code changes in Bitcoin Core, but in every other software that checks the blockchain POW, e.g. btcd.

jtimon replied:
well, they will need changes to support chains whose genesis block doesn't passs pow anyway, but yeah

practicalswift commented on src/primitives/block.h:
```diff
@@ -116,6 +116,50 @@ class CBlock : public CBlockHeader
+     *
+     * This operation may fail and return false, if no witness commitment exists upon call time. Returns true on success.
+     */
+    bool SetWitnessCommitmentSection(const uint8_t header[4], const std::vector<uint8_t> data);
```
`data` could be const ref?

kallewoof replied:
I think I meant to do that, yeah. Thanks, fixed!

practicalswift commented on src/primitives/block.h:
```diff
@@ -116,6 +116,50 @@ class CBlock : public CBlockHeader
+    /**
+     * The tx based equivalent of the above.
+     */
+    static bool SetWitnessCommitmentSection(CMutableTransaction& tx, const uint8_t header[4], const std::vector<uint8_t> data);
```
`data` could be const ref?

practicalswift commented on src/chainparams.cpp:
```diff
@@ -246,13 +266,101 @@ class CTestNetParams : public CChainParams {
+ */
+class SigNetParams : public CChainParams {
+public:
+    SigNetParams(const ArgsManager& args) {
```
Single parameter ctor could be `explicit`?

practicalswift commented on src/primitives/block.cpp:
```diff
@@ -29,3 +29,65 @@ std::string CBlock::ToString() const
+    return false;
+}
+
+bool CBlock::SetWitnessCommitmentSection(CMutableTransaction& mtx, const uint8_t header[4], const std::vector<uint8_t> data)
```
`data` const ref?

practicalswift commented on src/primitives/block.cpp:
```diff
@@ -29,3 +29,65 @@ std::string CBlock::ToString() const
+    return true;
+}
+
+bool CBlock::SetWitnessCommitmentSection(const uint8_t header[4], const std::vector<uint8_t> data)
```
Same here :)

practicalswift commented on src/rpc/mining.cpp:
```diff
@@ -948,6 +964,107 @@ static UniValue estimaterawfee(const JSONRPCRequest& request)
+    }
+
+    auto block = BlockAssembler(Params()).CreateNewBlock(coinbase_script)->block;
+    unsigned int extra_nonce = 0;
```
The scope for `extra_nonce` could be reduced.

kallewoof replied:
It doesn't feel cleaner to put it inside the lock brackets as those are meant to do the lock, not limit scope. Unrelated, but I am considering switching to using `GetTimeMillis()` instead of `0`, to make multiple calls to grindblock actually get different results. Thoughts welcome, since your eyes were here recently. :)

practicalswift commented on src/wallet/rpcwallet.cpp:
```diff
@@ -4160,6 +4162,80 @@ UniValue walletcreatefundedpsbt(const JSONRPCRequest& request)
+
+    CScript blockscript(g_signet_blockscript.begin(), g_signet_blockscript.end());
+
+    std::vector<uint8_t> sigdata;
```
`sigdata` not used? :)

practicalswift commented on contrib/signet/addtxtoblock.py:
```diff
@@ -0,0 +1,58 @@
+                                        epilog='''Help text and arguments:''',
+                                        formatter_class=argparse.RawTextHelpFormatter)
+    # parser.add_argument('--filter', help='filter scripts to run by regular expression')
+    args, unknown_args = parser.parse_known_args()
```
Idiomatic alternative which signals that `args` is intentionally unused: `_, unknown_args = parser.parse_known_args()` :)

practicalswift commented on test/functional/mining_signet.py:
```diff
@@ -0,0 +1,95 @@
+        self.skip_if_no_wallet()
+
+    def generate(self, node, count):
+        for i in range(count):
```
Nit: Same here: could use `_`.

practicalswift commented on test/functional/mining_signet.py:
```diff
@@ -0,0 +1,95 @@
+    def skip_test_if_missing_module(self):
+        self.skip_if_no_wallet()
+
+    def generate(self, node, count):
```
Nit: This could be a function instead of a method.

practicalswift commented on test/functional/mining_signet.py:
```diff
@@ -0,0 +1,95 @@
+        # give the privkey to node 1 so it can sign
+        node.importprivkey(private_key)
+        self.log.info('Imported network private key')
+        self.log.info('address: %s, privkey: %s' % (address, node.dumpprivkey(address)))
```
Instead of string format arguments use logging function parameters when using `self.log.info` :)

kallewoof replied:
You mean ` % (...)` → `, ...`? OK.

carnhofdaki commented:
No matter how far in syncing signet, it always says `progress=1.000000`.

```
...
2019-10-08T14:32:08Z UpdateTip: new best=00000098714cd57f96b1bb4bdec02a869cb3269351bb2e110d56ea480e9a9e41 height=3140 version=0x30000000 log2_work=32.087826 tx=3152 date='2019-09-09T00:13:04Z' progress=1.000000 cache=0.4MiB(3155txo)
2019-10-08T14:32:08Z UpdateTip: new best=0000014e412d4d959c4d397cdd685d66cd32df7842984f490d493f57cdd031ac height=3141 version=0x30000000 log2_work=32.088886 tx=3153 date='2019-09-09T00:22:06Z' progress=1.000000 cache=0.4MiB(3156txo)
2019-10-08T14:32:08Z UpdateTip: new best=0000031045f7a9f625ea845ace16fe18153bbcfaf2445951df96c37d427a34a5 height=3142 version=0x30000000 log2_work=32.089945 tx=3154 date='2019-09-09T00:31:07Z' progress=1.000000 cache=0.4MiB(3157txo)
2019-10-08T14:32:08Z UpdateTip: new best=0000044ee1b6e1b471d28f3f32e4c92deb9828efc2ab5948e76ad70c753ad7c8 height=3143 version=0x30000000 log2_work=32.091004 tx=3155 date='2019-09-09T00:40:09Z' progress=1.000000 cache=0.4MiB(3158txo)
2019-10-08T14:32:08Z UpdateTip: new best=000003c670e58d4876da8ff3f21d12efd70984e30eb7f69e74c8a7754e5f4657 height=3144 version=0x30000000 log2_work=32.092062 tx=3156 date='2019-09-09T00:49:12Z' progress=1.000000 cache=0.4MiB(3159txo)
```

I do everything according to Getting Started section of https://en.bitcoin.it/wiki/Signet#Fetch_and_compile_signet

carnhofdaki commented:
Also, errors like this are being shown on the console (yes, I am not running as daemon, but rather in `tmux`):

```
2019-10-08T15:13:23Z ERROR: AcceptBlockHeader: prev block not found
2019-10-08T15:13:26Z ERROR: AcceptBlockHeader: prev block not found
```

Let me know if it is proper that I am writing it here in this pull request.

jtimon commented on src/chainparams.cpp:
```diff
@@ -38,6 +40,12 @@ static CBlock CreateGenesisBlock(const char* pszTimestamp, const CScript& genesi
     return genesis;
 }
 
+static inline CBlock CreateGenesisBlock(const char* pszTimestamp, const CScript& genesisOutputScript, uint32_t nTime, uint32_t nNonce, uint32_t nBits, int32_t nVersion, const CAmount& genesisReward)
```
This is only called from the next CreateGenesisBlock, there's no need for a new function.

jtimon commented on src/chainparams.cpp:
```diff
@@ -49,20 +57,31 @@ static CBlock CreateGenesisBlock(const char* pszTimestamp, const CScript& genesi
+CBlock CreateSignetGenesisBlock(const CScript& block_script, uint32_t block_nonce)
+{
+    CHashWriter h(SER_DISK, 0);
+    h << block_script;
```
Sorry to reiterate this. If it was answered elsewhere I missed it.
Could we add the following?

```
    h << strNetworkID;
```

I know it requires a network restart for the current signet network, but I think that should be alright to restart the network until the PR is merged.
Even if we keep mining genesis blocks, this should make it easier to create a new signet chain with a different genesis block hash, even if the -signet_blockscript is the same.

kallewoof replied:
I agree adding network ID is a good idea. I'd like to not restart the network until things are more finalized, but I will put this on the list of things to change for the next reset.

jtimon commented on src/chainparamsbase.cpp:
```diff
@@ -23,6 +24,12 @@ void SetupChainParamsBaseOptions()
+    gArgs.AddArg("-signet_genesisnonce", "Nonce value for use in genesis block to satisfy proof of work requirement", ArgsManager::ALLOW_INT, OptionsCategory::CHAINPARAMS);
+    gArgs.AddArg("-signet_hrp", "Human readable part of bech32 address (suffixed by \"sb\"; default = \"\")", ArgsManager::ALLOW_ANY, OptionsCategory::CHAINPARAMS);
+    gArgs.AddArg("-signet_enforcescript", "Blocks must satisfy the given script to be considered valid (this replaces -signet_blockscript, and is used for opt-in-reorg mode)", ArgsManager::ALLOW_STRING, OptionsCategory::CHAINPARAMS);
+    gArgs.AddArg("-signet_seednode", "Specify a seed node for the signet network (may be used multiple times to specify multiple seed nodes)", ArgsManager::ALLOW_STRING, OptionsCategory::CHAINPARAMS);
```
The hrp and seeds arguments aren't signet specific necessarily. You could drop the 'signet' prefix and make them configurable for regtest too.
See https://github.com/bitcoin/bitcoin/pull/17037 for potential synergies.
You could also make SigNetParams extend CRegTestParams for more code reusability.

kallewoof replied:
That seems sensible in #17037, but there is literally no reason outside signet for these, here. I don't think people will be too upset if we rename these a few months after merge, if worse comes to worst.

jtimon commented on src/qt/networkstyle.cpp:
```diff
@@ -19,7 +19,8 @@ static const struct {
     {"main", QAPP_APP_NAME_DEFAULT, 0, 0},
     {"test", QAPP_APP_NAME_TESTNET, 70, 30},
-    {"regtest", QAPP_APP_NAME_REGTEST, 160, 30}
+    {"regtest", QAPP_APP_NAME_REGTEST, 160, 30},
```
I forgot this ```,```, this +1-1 is my fault. My apologies.

jtimon commented on src/chainparams.cpp:
```diff
@@ -248,13 +268,101 @@ class CTestNetParams : public CChainParams {
+        nPruneAfterHeight = 1000;
+
+        genesis = CreateSignetGenesisBlock(g_signet_blockscript, genesis_nonce);
+        consensus.hashGenesisBlock = genesis.GetHash();
```
why not add the following here?
```
        consensus.hashGenesisBlock = genesis.GetHash();
        checkpointData = {
            {
                {0, uint256S(consensus.hashGenesisBlock.GetHex())},
            }
        };

```

kallewoof replied:
Does it change anything? (Actual question; I assumed it didn't, except for blocks > height 0)

jtimon replied:
Well, I was thinking more on consistency with regtest, but functionally, I'm actually not sure it does anything (here or in regtest) either.

carnhofdaki commented:
> Also, errors like this are being shown on the console (yes, I am not running as daemon, but rather in `tmux`):
> 
> ```
> 2019-10-08T15:13:23Z ERROR: AcceptBlockHeader: prev block not found
> 2019-10-08T15:13:26Z ERROR: AcceptBlockHeader: prev block not found
> ```
> 
> Let me know if it is proper that I am writing it here in this pull request.

Not on current master. I am using https://github.com/carnhofdaki/bitcoin/tree/jt-signet-2-rebased-to-master

carnhofdaki commented:
@kallewoof is signet meant to overwrite file `blk00000.dat` and not have any index? How is the explorer at https://explorer.bc-2.jp possible then?

kallewoof commented:
@carnhofdaki Thanks for testing! Comments below.

> No matter how far in syncing signet, it always says `progress=1.000000`.

I'm confused. My bitcoin full node running on mainnet has this
```
2019-10-10T04:59:39Z UpdateTip: new best=0000000000000000000c41823b84f6273e2bf59648cfa7187eb2919213643a3d height=598693 version=0x20800000 log2_work=91.193405 tx=463424821 date='2019-10-10T04:58:54Z' progress=1.000000 cache=21.4MiB(60544txo) warning='45 of last 100 blocks have unexpected version'
```
so the behavior seems ordinary to me. What's the problem?

> 2019-10-08T15:13:23Z ERROR: AcceptBlockHeader: prev block not found

This is because there are a couple of other signets running that have not properly removed the seed node parameter, so they are trying to connect to incompatible nodes (their genesis block is different, so in the end they get disconnected).

It's a nuisance, but not a big problem.

> is signet meant to overwrite file blk00000.dat and not have any index? How is the explorer at https://explorer.bc-2.jp possible then?

SIgnet works exactly like mainnet/testnet/regtest in regards to file I/O. By index, do you mean a transaction index? You need to do txindex=1 in your bitcoin.conf for that to appear. If you did, and it still isn't appearing, do let me know. :)

kallewoof commented:
@jtimon Thanks for all the feedback/suggestions. I'm trying to work through them, but am a bit swamped at work right now so it may be a bit slower than normal. Will get back to you though.

carnhofdaki commented:
@kallewoof You are welcome! See my comments below.

> I'm confused. My bitcoin full node running on mainnet has this
> 
> ```
> 2019-10-10T04:59:39Z UpdateTip: new best=0000000000000000000c41823b84f6273e2bf59648cfa7187eb2919213643a3d height=598693 version=0x20800000 log2_work=91.193405 tx=463424821 date='2019-10-10T04:58:54Z' progress=1.000000 cache=21.4MiB(60544txo) warning='45 of last 100 blocks have unexpected version'
> ```
> 
> so the behavior seems ordinary to me. What's the problem?

What you are showing is an actually synced bitcoin mainnet node. When you start syncing
bitcoin mainnet without having any blocks, it will start with `progress=0.000000` and slowly
add there according to block index number devided by total blocks currently (and time, this
is not an exact formula, just to get an idea).

> > 2019-10-08T15:13:23Z ERROR: AcceptBlockHeader: prev block not found
> 
> This is because there are a couple of other signets running that have not properly removed the seed node parameter, so they are trying to connect to incompatible nodes (their genesis block is different, so in the end they get disconnected).
> 
> It's a nuisance, but not a big problem.

I see.

> > is signet meant to overwrite file blk00000.dat and not have any index? How is the explorer at https://explorer.bc-2.jp possible then?
> 
> SIgnet works exactly like mainnet/testnet/regtest in regards to file I/O. By index, do you mean a transaction index? You need to do txindex=1 in your bitcoin.conf for that to appear. If you did, and it still isn't appearing, do let me know. :)

I will re-run with `txindex=1` but what I mean here is that when I run a signet full node, even after
full sync to the block 8002, I have only file `blk00000.dat` with has last modification date set to
when the last block was received.

carnhofdaki commented:
@kallewoof So I was playing more.

```
#!/bin/sh
while true
  do ./bitcoin-cli -datadir=signet \
    sendtoaddress `./sn-na.sh` `./sn-gb.sh` "" true true
  sleep 1
done
```

Where `sn-na.sh` and `sn-gb.sh` are respectively:

```
./bitcoin-cli -datadir=signet getnewaddress
```

and

```
./bitcoin-cli -datadir=signet getbalance
```

See this anomaly - it does not add up: https://explorer.bc-2.jp/tx/7bc98cb0298233bbeb8d43deebc879626c3403add1ebbf443c44a31576283aff

I hope that is [esplora](https://github.com/Blockstream/esplora)'s bug

I have run it multiple times and the output was:

```
# before this I did many one-liner transactions
2762756521b02c51eb403f8b82b8f196775af1b6fd5487f7b9734d2ff3efa4e3
8def17c72527015c22320cd501bbeac1230f169a21e4bbfde526c6fbf25ae373
f72260cf97ca06ca836cf814b2413c784d21b75613fecdb6c1a5f937b53d79f0
4ff2e73d8c5dff14e10d63dd995634f4cc23968aa1e45fbaeff0590ddb23e632
e9e61b32f7c449f0e52cfbb0d74981d87ba83f116186f39b1e839d8dfc4703c1
error code: -3
error message:
Invalid amount for send
error code: -3
error message:
Invalid amount for send
error code: -3
error message:
Invalid amount for send
^C
```

Then next time:

```
aa9c48e3a761585d441e6b0bb5d36fc352e0c027e92e1f2f746f078c0647f807
c2cdc6e522e226bed519c0c3480a3077e6de94032cc1dd6039094589bed2f591
54a802861bd40b18f4307583bd33c27a1dae8fae90a685e355183e9ac331d28d
b154ee4344da0eb093d72f8f45122ccdbc3736799c49251fed0694ff014cee4b
951ed2d88878b0daac4ea9e1d8e5a7d8175af43305f357ac87e315676c48221a
99a4e911bd02d1ba9bc15cf980f6680a3035c48012d572262b677de04aa76fa4
23c51dcbe708d9e7a065bc66cd8f8d4b17fe205664eab4a6b3fa25196d1d9bec
a838decadcfa470e32d5016d9dd66d3445c12596b513b518f37d6813fa5d492f
1e10d06a53c43781172971e3af5bc2e90ca8a8618c324c945d87d3a9ffc4e001
03a645782b829916298dd89a6767f5a056178f6ff2af3aa6fb228c218c870f2f
425da2e23eaf93efa3823bc95b37563fadc11f5d57a5411ca58e24c0b066bc6a
8e9b912a2695f7c70e06fe4d66c972cf3174cd6bfe551f36bf712b32bb0adc87
09ffd34984f445c5bbd546a80fa268f8922884e3aee6eefcc63115de43da0255
0a376080f6d1611dcc22907575e57e21fb04b8b4e14a58148e0ea8ac6d848772
e7da0a599288e5772c382b19981119bfb19f01c1c06fb885edb2e259246568b2
ccef452a96821edae47fd715d6869236240f5fbe93ed5af0fec9717807949b43
d73eed9ebba93fae7efcdc3b482292aa7c1405be0b21eb7fe6f0f9105c38ba1a
c9a080b5c316103524fbaa3d24ef892c63f0cfb2769fa0e70e6e116140d2db8b
06eea69aedddbd87de52da326a67f0688ec4b628dd049715a39490afce8a4845
55dbb48b94b7f387cea294160c601517321dc4271fd638159b8cfe8f4f097153
7bc98cb0298233bbeb8d43deebc879626c3403add1ebbf443c44a31576283aff
b84e6d3d24f9e776af0e7c63df4d99280e325bb37a6603b6af73ed137206053e
53de67fb023c9bb3d4d9f1390f122272246270bfe469ef162a86f8202a918b54
140524494767498a07a2ea316b4880af6b872cb49578cf60d6205ce4e674e7b5
5560a252a6ed71f63fc2d87bc7c95c459c2586553b92a76abe6630ad8f6f0a0e
ee4ac451e680d53961dc8030d7621ff705af50f46431b0a0ef32d3fabd739824
error code: -3
error message:
Invalid amount for send
error code: -3
error message:
Invalid amount for send
^C
```

See block [8174](https://explorer.bc-2.jp/block/0000021f73ceda899d47c05288e2dbab83d6a79b3ec15f3f24730bf8075ad184)

carnhofdaki commented:
Testing on c584956573b553f33771b7d853e2ccf8f033d29a, ARM, Ubuntu 19.04, but berkleydb and boost are built from `bitcoin/depends` and installed to `/usr` (though not related to my issues).

I have issues with running a [custom signet](https://en.bitcoin.it/wiki/Signet#Custom_Signet) always ending on the last step - generating a block and signing it where I get:

```
error code: -25
error message:
could not produce a signature -- do you have the private key(s)?
```

And everytime (including the first time) I run `importprivkey $PRIVKEY` I get following in the daemon log:

```
2019-10-11T11:29:46Z [default wallet] Already have key with pubkey 0353135RED5871a7, skipping                  
2019-10-11T11:29:46Z [default wallet] Already have script 0014046RED3a5ede1, skipping
```

carnhofdaki commented:
> Concept ACK, though I am unsure if signet should be configurable at all. (via `signet_blockscript` for example)
> 
> It should just be a shared public network, an alternative (or replacement) of testnet.

@NicolasDorier as of my testing, the configurable signet currently does not work. I will be happy if someone can prove me wrong, but the steps at [Signet - Bitcoin Wiki](https://en.bitcoin.it/wiki/Signet#Custom_Signet) currently do not lead me to a working custom signet.

MarcoFalke commented on src/primitives/block.h:
```diff
@@ -116,6 +116,50 @@ class CBlock : public CBlockHeader
+     *
+     * Returns -1 if no witness commitment was found.
+     */
+    template<typename T> static inline int GetWitnessCommitmentIndex(const T& coinbase) {
```
Why is this moved? I think the goal for `primitives` is to be data-only and don't contain any helper logic, especially if they are consensus critical.

kallewoof replied:
It seems very weird to keep this obviously `CBlock`-related functionality in some completely different place. I'm not sure about primitives and data-only logic. If people feel strongly, I can restore this. Should the other stuff (witness commitment section stuff) follow along into validation.cpp then, you think?

kallewoof commented:
@carnhofdaki Can you make a log file showing you all the steps of you setting up a custom signet to you getting the error? (and accompanying debug.log file would be helpful too)

carnhofdaki commented:
@kallewoof sure, on it now

kallewoof commented:
@carnhofdaki Sorry for late response. Tried your script and, yeah, that looks off. Will have to look closer into what might be causing it.

carnhofdaki commented:
> @carnhofdaki Sorry for late response. Tried your script and, yeah, that looks off. Will have to look closer into what might be causing it.

Thank you for having a look at the script!

kallewoof commented:
@carnesen Sorry for delay on looking at that script, work is keeping me away from everything atm. Hoping to get back into things in a week or two, though. Will look then.

carnesen commented:
> @carnesen Sorry for delay on looking at that script, work is keeping me away from everything atm. Hoping to get back into things in a week or two, though. Will look then.

@carnhofdaki I think this was meant for you (cc @kallewoof)

kallewoof commented:
@carnesen Thanks. Yeah, it was.

fjahr commented:
Concept ACK

Great work! I set up a custom signet a few months ago and found it very easy to do and it was running smoothly (followed the wiki article, not a review, code has changed a lot since then). However, I also discovered that (as is usually the case) coordinating other participants in that Signet was the hard part. So, I now think the main contribution of Signet will be the global signet with regularly scheduled fee spikes, attack scenarios etc. where the behavior of all kinds of upcoming PRs can be tested and measured on a regular basis, especially P2P behavior. But even if custom signets may be used less than I thought initially, I still see the value in them for upcoming softforks like taproot. Will try to review the code in the coming weeks.

practicalswift commented:
This PR has gotten four Concept ACKs and zero Concept NACKs if I'm counting correctly.

Please chime in: more Concept ACKs or Concept NACKs?

The benefits have been discussed extensively: what costs can we see?

Personally I think signet would be very valuable to have when testing.

kallewoof commented:
@carnhofdaki 

> @kallewoof See log and the script which produced it at https://gist.github.com/carnhofdaki/60edef577f637ef2dbf4d244e4e279c2
> 
> Feel free to ask any questions if you wonder about the commands in the script. Generally it is just following the [wiki article](https://en.bitcoin.it/wiki/Signet#Custom_Signet) but fully automated.

Sorry for the extremely late reply. I found the issue -- you are setting `LEN` to the length of the hexadecimal string, rather than the length of the actual data. One fix:
```diff
- LEN=$(printf $PUBKEY | wc -c)
+ LENX2=$(printf $PUBKEY | wc -c)
+ LEN=$((LENX2/2))
```

kallewoof commented:
@fjahr Thanks for the feedback! I was honestly dragged away from everything for the last couple of months, and am finally getting back to things now. Scheduling events on the global running signet is something I've been meaning to do for a long time. The code to do things like intentional reorgs and such is in place, but lacking a UI right now.

If you have any concrete ideas I'd love to hear them :)

@practicalswift Thanks for the nudge! I will bring up bumping signet over to high priority from needs concept ACK on IRC.

fjahr commented:
> If you have any concrete ideas I'd love to hear them :)

I am not sure about priorities as I am not working on something where I need it right now, but some ideas could be interesting:
- "larger" re-orgs
- large numbers of transactions with rising fee rate, so that default sized mempools start evicting transactions
- mining empty blocks
- full blocks but transactions were not in the mempool
- network hash-rate drop/spike

Ping @amitiuttarwar what would be helpful for your work on https://github.com/bitcoin/bitcoin/pull/16698?

kallewoof commented:
I have no idea why tests are failing... they aren't, locally.

MarcoFalke commented:
re-run ci

carnhofdaki commented:
@kallewoof Thank you! https://github.com/bitcoin/bitcoin/pull/16411#issuecomment-569483123
It works well now!

kallewoof commented:
I think we can remove "Needs Conceptual Review" tag as well.

kallewoof commented:
Bumped copyright year. Note that Travis tests did not run, only Appveyor.

jtimon commented:
Can we have the discussion on not checking the genesis' block pow vs needing grind the genesis block again?
Can anybody remind me what are the main arguments in favor of having to grind the genesis block?

kallewoof commented:
@jtimon I think we should discuss on the mailing list. I'll send an email listing pros and cons.

NicolasDorier commented:
I started talking with @kallewoof about signet, and I thought of something.

## Problem

Software built on top of bitcoin core very often hard code and rely on the genesis block hash.
However it seems that because of [this](https://github.com/bitcoin/bitcoin/pull/16411/files#diff-64cbe1ad5465e13bc59ee8bb6f3de2e7L26), each signet will have a different genesis block. 
This mean that every times you use a different signet, you would need to rebuild every software on top of of bitcoin core. This is a friction that would prevent people from using signet.

## Specific use case

For example, if a project using NBitcoin would decide to use another signet, the developer would need me to hardcode his new signet in my library then I need to push a new version of my library, then he would need to recompile his project to use this version and finally able to use the new signet.
Moving back and forth between signet would require such kind of heavy coordination between many people involved.

Because in real code the genesis block never change, it is not wise either for NBitcoin to allow dynamic network parameters. This would push the complexity to the user of my library, and make his code a bit farther away from their mainnet.

## Proposal

Instead of putting the challenge scriptPubKey in the genesis block, it should be put in the block just after.

The other advantage of it is that we don't need to grind the genesis block to pass the PoW check everytimes somebody change the signet params.

kallewoof commented:
It would definitely simplify software since most stuff I've found has a hard coded genesis block.

The downside is that most light clients (anything that *doesn't* also check the block signature) will run the risk of switching to the wrong chain!

kallewoof commented:
One partial solution to the problem would be to set the magic number (`pchMessageStart`) to be e.g. the first 4 bytes of `sha256d(signet_challenge)`. There would be a 1 in 2^32 chance that two signets' magic numbers would collide, but I think this is acceptable, as most SPV nodes will have trusted nodes they connect to for signet testing anyway.

Interested to hear people's thoughts!

Sjors commented:
It sounds reasonable to have a fixed genesis block for all signets, with differentiation happening in the first block. Taking the thought a bit further, could you have an arbitrary height where the signet constraint activates? That might make it easier to fork signets, so you can try new consensus features on an existing UTXO set.

The `sha256d(signet_challenge)` also sounds reasonable. You could write a test case with two different signets to demonstrate this.

jtimon commented:
> Instead of putting the challenge scriptPubKey in the genesis block, it should be put in the block just after.

But you need them to have different genesis blocks. For example, the genesis block is the chain_id in lightning networks.
See, for example, my project in which I use 3 different regtest chains for lightning.
I'm not using signet there, but it would be similar.

https://github.com/jtimon/multi-ln-demo/blob/master/multiln/chains.py

I guess that file is not very interesting, but different chains need different genesis blocks.

> This mean that every times you use a different signet, you would need to rebuild every software on top of of bitcoin core.

Alternatively those projects can reimplement how to calculate the genesis hash from the challenge script.
This alternative requires that the pow for genesis blocks isn't checked so that there's no need to grind the genesis block, which is desirable for simplicity anyway, and I am yet to hear a good argument against.

0xB10C commented:
Syncing signet (both the default one and a custom one) reports an incorrect IBD _progress_ while in IBD and after the IBD. The wrong progress can be seen in the debug.log and via `getblockchaininfo` and potentially causes confusion. Haven't looked into why the calculation is wrong.

```bash
./bitcoin-cli -signet getblockchaininfo
{
  "chain": "signet",
  "blocks": 27004,
  "headers": 27004,
  "bestblockhash": "0000015194fa56d07f9cf8ce3c23e3e5ab64d68cfa06fa15e622b86336db10c0",
  "difficulty": 0.002338098747362749,
  "mediantime": 1581068861,
  "verificationprogress": 5.484926296056364e-05,
  "initialblockdownload": false,
  "chainwork": "000000000000000000000000000000000000000000000000000000236405f47a",
  "size_on_disk": 10652880,
  "pruned": false,
  "signet-blockscript": "512103ad5e0edad18cb1f0fc0d28a3d4f1f3e445640337489abb10404f2d1e086be43051ae",
  "softforks": {
  // shortend  
  },
  "warnings": "This is a pre-release test build - use at your own risk - do not use for mining or merchant applications"
}
```

kallewoof commented:
@0xB10C Unfortunately this is not trivial to fix, because it depends on statistics manually put into the chainparams file (see e.g. the mainnet class a few pages up, where it has info about the number of transactions and such in the network). For a very-long-term signet, it might make sense to start adding those, but whenever the signet network switches to a new one, all of that would become redundant.

ajtowns commented:
> > 
> 
> But you need them to have different genesis blocks. For example, the genesis block is the chain_id in lightning networks.

@jtimon per [bolt-00](https://github.com/lightningnetwork/lightning-rfc/blob/master/00-introduction.md#glossary-and-terminology-guide) you don't have to use the genesis block as the chain id, so I think it should be fine to use block 1 as the chain id for testing lightning on signet

kallewoof committed:
add signet chain and accompanying parameters

kallewoof committed:
rpc: add grindblock command for generating POW on a signed signet block

kallewoof committed:
rpc: refactor out grindBlock part from generateBlocks

kallewoof committed:
add simple signature support (checker)

kallewoof committed:
add simple signature support (creator)

jtimon commented:
> > > 
> > 
> > 
> > But you need them to have different genesis blocks. For example, the genesis block is the chain_id in lightning networks.
> 
> @jtimon per [bolt-00](https://github.com/lightningnetwork/lightning-rfc/blob/master/00-introduction.md#glossary-and-terminology-guide) you don't have to use the genesis block as the chain id, so I think it should be fine to use block 1 as the chain id for testing lightning on signet

Well, but in that case, you would need to hardcode the second block instead of the genesis block and you have the same problem @NicolasDorier complains about, no?

I understand that some use cases don't need a chain_id or anything like that, and for those it is fine to have the same the same genesis hash and they don't even care about the second block hash because they won't hardcode that either.
But for lightning, I think you need a chain_id.

Supporting multiple chains doesn't require external software to hardcode every new chain though, that's just a design decision.
Just like bitcoin core generates the hash in a deterministic way, other software can too.
Instead of implementing only a few hardcoded signets or only the default one, external software could implement them all possible signets at once, just like bitcoin core does.

For this, it would help to not mine the genesis block and thus not need to handle the nonce as configurable anywhere, but that's orthogonal.
Sorry to insist, but can we discuss what are the advantages of grinding the genesis block again?

Back to @NicolasDorier 's proposal of committing the challenge to the second block instead of the first one, I'm all fine with that, it's just that I also want to generate chains that do have different genesis blocks when I want to. This could be done, for example, by hashing the network name.
People wanting the same genesis hash could all just leave the default "signet" while others could customize different names, resulting in different genesis blocks, like in https://github.com/bitcoin/bitcoin/pull/17037

This again would be simpler if the genesis block wasn't mined.

Or perhaps we could even allow both having the challenge being committed on the genesis block or on the second one as a configuration option?

kallewoof commented:
I just want to add to this that, with the hard coded genesis block, all software will still need to re-generate the magic number (message header / message start / whatever you wanna call it) when they switch signets. I think this is significantly easier than a dynamic genesis block hash, but it's still not "plug-and-play" to switch to a different signet. Ping @NicolasDorier.

kallewoof commented:
Closing in favor of #18267.
